### DISCUSSION TITLE
Is btape still working with 23.0.5 and LTO9 tapes?

### DISCUSSION THREAD

**Discussion by kabassanov** [bareos/bareos#1997] [General]
Hi,
I try to validate btape fill tests with my LTO9 drives (and autochanger), but every multivolume test fails after reposition to block 0:1 while reading the second tape (the first one is correctly written and read) :

06:58:06 Flush block, write EOF
Wrote block=17000000, file,blk=17839,385 VolBytes=17,825,790,951,424 rate=121.7 MB/s
Wrote block=17005000, file,blk=17844,620 VolBytes=17,831,033,831,424 rate=121.6 MB/s
31-Oct 06:59 btape JobId 0: End of Volume "TestVolume1" at 17844:713 on device "Drive-2" (/dev/nst/tape1). Write of 1048576 bytes got -1.
31-Oct 06:59 btape JobId 0: Re-read of last block succeeded.
btape: stored/btape.cc:2643-0 Last block at: 17844:712 this_dev_block_num=713
btape: stored/btape.cc:2672-0 End of tape 17844:0. Volume Bytes=17,831,131,348,992. Write rate = 121.6 MB/s
31-Oct 06:59 btape JobId 0: End of medium on Volume "TestVolume1" Bytes=17,831,131,348,992 Blocks=17,005,092 at 31-Oct-2024 06:59.
31-Oct 06:59 btape JobId 0: 3307 Issuing autochanger "unload slot 1, drive 1" command.
...
31-Oct 07:05 btape JobId 0: 3304 Issuing autochanger "load slot 2, drive 1" command.
31-Oct 07:05 btape JobId 0: 3305 Autochanger "load slot 2, drive 1", status is OK.
Wrote Volume label for volume "TestVolume2".
31-Oct 07:06 btape JobId 0: Moving to end of data on volume "TestVolume2"
31-Oct 07:06 btape JobId 0: Warning: For Volume "TestVolume2":
The number of files mismatch! Volume=1 Catalog=0
Correcting Catalog
31-Oct 07:06 btape JobId 0: New volume "TestVolume2" mounted on device "Drive-2" (/dev/nst/tape1) at 31-Oct-2024 07:06.
btape: stored/btape.cc:2243-0 Wrote 1000 blocks on second tape. Done.
Done writing 0 records ...
Wrote End of Session label.
btape: stored/btape.cc:2310-0 Wrote state file last_block_num1=712 last_block_num2=47
btape: stored/btape.cc:2331-0

07:06:22 Done filling tapes at 2:49. Now beginning re-read of first tape ...
btape: stored/btape.cc:2413-0 Enter do_unfill
31-Oct 07:06 btape JobId 0: 3307 Issuing autochanger "unload slot 2, drive 1" command.
31-Oct 07:07 btape JobId 0: 3304 Issuing autochanger "load slot 1, drive 1" command.
31-Oct 07:09 btape JobId 0: 3305 Autochanger "load slot 1, drive 1", status is OK.
31-Oct 07:09 btape JobId 0: Ready to read from volume "TestVolume1" on device "Drive-2" (/dev/nst/tape1).
Rewinding.
Reading the first 10'000 records from 0:0.
10'000 records read now at 1:313
Reposition from 1:313 to 17844:712
Reading block 712.

The last block of the first tape matches.

31-Oct 07:10 btape JobId 0: 3307 Issuing autochanger "unload slot 1, drive 1" command.
31-Oct 07:12 btape JobId 0: 3304 Issuing autochanger "load slot 2, drive 1" command.
31-Oct 07:12 btape JobId 0: 3305 Autochanger "load slot 2, drive 1", status is OK.
31-Oct 07:13 btape JobId 0: Ready to read from volume "TestVolume2" on device "Drive-2" (/dev/nst/tape1).
Reposition from 0:0 to 0:1
Reading block 1.
Error reading block: ERR=stored/block.cc:1025 Read zero bytes at 0:0 on device "Drive-2" (/dev/nst/tape1).

btape: stored/btape.cc:2340-0 do_unfill failed.

I don't really understand what this means... In particular why there is a repositionning to 0:1 and why a read operation is tried on 0:0 ?
What do I understand in a wrong manner?

Another question is why if itdt tool reports 300MB/s without compression, I can only get 121 - 132 MB/s with btape (an regular writing job)?

Thanks.

**Comment by arogge** [comment]
My best guess right now is we broke that in #1979, I'll have to check that.

**Reply by kabassanov** [reply]
So, it is much more a storage daemon issue than btape issue? That sugguests that every multi volumes task will fail...

**Reply by arogge** [reply]
Yes and no. As far as I can tell right now the issue is that btape was not adapted to a rather recent change in the storage daemon code.
However, I'm still not finished identifying the problem, so currently it is just a hypothesis.
Normal jobs (i.e. everything not using btape) should work fine (and works fine with our testsuite)

**Reply by arogge** [reply]
So I can now confirm that btape (and only btape) is broken. As I thought we broke it in #1979.
The problem is that btape repositions to 0:1 (i.e. the second block in the first tape-file), but should reposition to 1:0 (i.e. the first block in the second tape-file).
I think it should not reposition in the first place, but actually read the volume label (which would have done the right thing).

**Reply by kabassanov** [reply]
Not sure to correctly understand what should happen during btape test...
Once tape1 has been filled, it loads the tape 2 and writes 1000 blocks (is a block = a record?).
Then it reloads tape 1 and reads 10000 records from 0:0, then goes to the last block of tape 1 and read it.
On success, tape 2 is loaded again, and it should read its 1000 blocks previously written from 0:0. Am I wrong? In this context, it is not clear for me what means the following:
 
btape: stored/btape.cc:2243-0 Wrote 1000 blocks on second tape. Done.
Done writing 0 records ...

I checked manually (with btape) that 0:1 is not readable on the second tape, but did not test reading from 1:0... As my test tapes have been ereased, I'd make the test once again and verify...

**Reply by arogge** [reply]
Just give me a day or two to come up with a fix.

**Reply by kabassanov** [reply]
No problem. Thanks!

**Reply by arogge** [reply]
Just to explain what is happening here:
Previously the beginning of the tape would look like this:
```
0:0           0:1          0:2
 +-------------+------------+
 | Label Block | Data Block |
 +-------------+------------+
```
With the slight change we introduced to the labeling process it now looks like this:
```
0:0           0:1 1:0          0:1
 +-------------+---+------------+
 | Label Block |EOF| Data Block |
 +-------------+---+------------+
```
When the tape driver sees that file mark (`EOF`) it will report zero bytes read. That's how you as a programmer are supposed to detect a file mark on the tape. After that you're in the next file at block 0 and can start to read it.

As `btape` still thinks the data starts at 0:1 it tries to read there, encounters the file mark, gets zero-length data and emits the error _"Read zero bytes at 0:0"_ (that should rather report _"Read zero bytes at 0:1"_).
The fix is to read the volume label instead of blindly repositioning to 0:1.

**Comment by arogge** [comment]
I just created pull request #2018 that addresses the issue.
Would you re-run your test once packages show up [here](https://download.bareos.org/experimental/PR-2018/)?

**Reply by kabassanov** [reply]
Hello,

It is working now (This time I managed tapes load/unload manually, because I did not associate this drive with the autochanger).
Thank you.

Multiple tape test selected.
21-Nov 09:34 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
btape: stored/btape.cc:2983-0 Autochanger returned: 0
Mount blank Volume on device "Drive-1" (/dev/nst/tape0) and press return when ready:
btape: stored/btape.cc:2989-0
Wrote Volume label for volume "TestVolume1".
Wrote Start of Session label.
09:34:37 Begin writing Bareos records to first tape ...
Wrote block=5000, file,blk=1,4999 VolBytes=5,241,831,424 rate=262.0 MB/s
...
Wrote block=15370000, file,blk=76,10074 VolBytes=16,116,612,071,424 rate=254.1 MB/s
22-Nov 03:11 btape JobId 0: End of Volume "TestVolume1" at 76:10335 on device "Drive-1" (/dev/nst/tape0). Write of 1048576 bytes got -1.
22-Nov 03:11 btape JobId 0: Re-read of last block succeeded.
btape: stored/btape.cc:2657-0 Last block at: 76:10334 this_dev_block_num=10335
btape: stored/btape.cc:2686-0 End of tape 76:0. Volume Bytes=16,116,885,749,760. Write rate = 254.0 MB/s
22-Nov 03:11 btape JobId 0: End of medium on Volume "TestVolume1" Bytes=16,116,885,749,760 Blocks=15,370,260 at 22-Nov-2024 03:11.
22-Nov 03:12 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
btape: stored/btape.cc:2983-0 Autochanger returned: 0
Mount blank Volume on device "Drive-1" (/dev/nst/tape0) and press return when ready:
btape: stored/btape.cc:2989-0
Wrote Volume label for volume "TestVolume2".
22-Nov 07:20 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
22-Nov 07:21 btape JobId 0: Moving to end of data on volume "TestVolume2"
22-Nov 07:21 btape JobId 0: Warning: For Volume "TestVolume2":
The number of files mismatch! Volume=1 Catalog=0
Correcting Catalog
22-Nov 07:21 btape JobId 0: New volume "TestVolume2" mounted on device "Drive-1" (/dev/nst/tape0) at 22-Nov-2024 07:21.
btape: stored/btape.cc:2250-0 Wrote 1000 blocks on second tape. Done.
Done writing 0 records ...
Wrote End of Session label.
btape: stored/btape.cc:2317-0 Wrote state file last_block_num1=10334 last_block_num2=1000
btape: stored/btape.cc:2338-0
07:21:17 Done filling tapes at 1:1002. Now beginning re-read of first tape ...
btape: stored/btape.cc:2419-0 Enter do_unfill
22-Nov 07:21 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
btape: stored/btape.cc:2456-0 Autochanger returned: 0
Mount first tape. Press enter when ready:
btape: stored/btape.cc:2459-0
22-Nov 07:25 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
22-Nov 07:26 btape JobId 0: Ready to read from volume "TestVolume1" on device "Drive-1" (/dev/nst/tape0).
Rewinding.
Reading the first 10'000 records from 0:0.
10'000 records read now at 1:313
Reposition from 1:313 to 76:10334
Reading block 10334.
The last block of the first tape matches.
22-Nov 07:27 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
btape: stored/btape.cc:2519-0 Autochanger returned: 0
Mount second tape. Press enter when ready:
btape: stored/btape.cc:2522-0
22-Nov 07:38 btape JobId 0: No "Changer Device" for "Drive-1" (/dev/nst/tape0). Manual load of Volume may be required.
22-Nov 07:38 btape JobId 0: Ready to read from volume "TestVolume2" on device "Drive-1" (/dev/nst/tape0).
Reposition from 0:0 to 1:0
Reading block 0.
The first block on the second tape matches.
Reposition from 1:1 to 1:1000
Reading block 1000.
The last block on the second tape matches. Test succeeded.
